core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -264,16 +290,26 @@ private[spark] class Executor(
     decommissioned = true
   }
 
+  private[executor] def createTaskRunner(context: ExecutorBackend,
+    taskDescription: TaskDescription) = new TaskRunner(context, taskDescription, plugins)
+
   def launchTask(context: ExecutorBackend, taskDescription: TaskDescription): Unit = {
-    val tr = new TaskRunner(context, taskDescription, plugins)
-    runningTasks.put(taskDescription.taskId, tr)
+    val taskId = taskDescription.taskId
+    val tr = createTaskRunner(context, taskDescription)
+    runningTasks.put(taskId, tr)
+    val killMark = killMarks.get(taskId)
+    if (killMark != null) {
+      tr.kill(killMark._1, killMark._2)
+      killMarks.remove(taskId)
+    }
     threadPool.execute(tr)
```

Comment:
If killed, do we need to execute it ?

Reply:
Yes. Otherwise we won't get the right status update.

Reply:
You can pull the `TaskKilledException` catch clause into a separate method and directly invoke it from here.

Reply:
It seems safer to me to go through the normal task execution code path, which deals with the case of "running an already killed task", see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/executor/Executor.scala#L471-L479

Reply:
In this context, we know it is going to immediately fail when the task starts - and there is a bunch of things which happen before that check and after here.
I am fine with adding this as a TODO if you feel strongly about keeping the current path.

Reply:
The argument of doing the kill inside the TaskRunner rather than here would be:
1. not to duplicate the code for the kill routine;
2. the kill routine can be executed in a separate thread.

Reply:
Ideally we should kill the task as early as possible, but in this case, I think it's safer to reuse the existing code path and launch an already killed task, instead of adding a new code path to skip task launching, which may miss some bookkeeping work or event triggering.